### DISCUSSION TITLE
More Advice Around Environments (Creating Them, Managing Them, etc) Needed

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#96] [Discussions]
My application is becoming quite complex and I'm facing the issue of how to manage environments. Up till now, I've gotten away with using a common interactive environment via `Eval()`, where my application interacts with it iteratively/incrementally, from diverse places in the code.

Now it's getting a bit harder to think about these interactions as my application's natural complexity is emerging. I was wondering if someone who knows (@leppie I assume, but anyone who has encountered this issue is welcome to chime in) the practicalities of managing multiple environments in a Windows app could provide a framework to think about this.

I am aware that loading libraries is a one-timed thing in an environment, even if they subsequently change. Since my application encourages users to create libraries, I will need a way for the environment to reload when libraries are changed, and I'm not sure how to approach this in the abstract, which is why I'm seeking some concrete ways to go about it. Yes, my app will encourage running one-off test scripts to facilitate creating new libraries, but as a practical matter it is undesirable when something does change to expect them to close and reopen my application.

Anyway... a kind of "taxonomy of environment management strategies" would be helpful here, if anyone has such a thing, or ideas how to frame it.

**Comment by leppie** [comment]
File based libraries
```
> (import (test))
> (main)
brainssssssssssssssssssssin
> (import (test))
> (main)
brainssssssssssssssssssssin
> (uninstall-library '(test))
> (main)
Unhandled exception during evaluation:
&who: main
&message: "identifier out of context"
&syntax:
  form: main
  subform: #f
&trace: #<syntax main>
;; I edit the file now
> (import (test))
> (main)
ssssssssssssssssssssin
>
```

Another way to create libraries is to define them directly in the REPL.
```
> (library (test)
.   (export main)
.   (import (rnrs))
. (define (main)
.   (display "ssssssssssssssssssssin\n")))
> (import (test))
> (main)
ssssssssssssssssssssin
> (library (test)
.   (export main)
.   (import (rnrs))
. (define (main)
.   (display "fffffffn\n")))
> (main)
Unhandled exception during evaluation:
&who: main
&message: "identifier out of context"
&syntax:
  form: main
  subform: #f
&trace: #<syntax main>

> (import (test))
> (main)
fffffffn
```

There is a parameter called `allow-library-redefinition` which is `#t`  for REPL. You might have to set that to get it to work when calling eval.

Also look at the `new-interaction-environment` procedure. It accepts an optional libspec ie `'(ironscheme)` arg. If not provided, `'(ironscheme)` is the default (as you can guess). 

You can then use this environment to pass to `Eval` on say a per user basis. But imported libraries are still going to be shared. Not sure what mileage you can get for your scenario.

**Reply by bcalco** [reply]
Would the libspec in the case of multiple libs look like this `'((ironscheme) (minikanren) ...)`?

**Reply by leppie** [reply]
Basically anything you can put inside `(import <import-spec>)`, but quoted.  See http://www.r6rs.org/final/html/r6rs/r6rs-Z-H-10.html

**Reply by bcalco** [reply]
What exactly are the semantics/consequences of `(allow-library-redefinition #t)`? Does it just mean `(uninstall-library ...)` actually removes the symbols (whereas otherwise it would be a no-op)? Does calling `(import (...))` now implicitly uninstall and reinstall the library?  

Oh, and do I define it in the toplevel `(interaction-environment)` once, or for each environment I create?

**Reply by leppie** [reply]
IIRC, `allow-library-redefinition` only works for libraries defined inside the REPL (iow not loaded from a file). It does not seem very useful now if only available in that context. Not sure if I can make it work for files (or maybe it does and I broke it?), but that would be useful. Will check it out.

**Reply by bcalco** [reply]
Yes, that would be great. The use case here obviously is rapid development in an application that can't reasonably be shut down and restarted for every change.

**Comment by bcalco** [comment]
Also, if present in the working directory, would `init.ss` be loaded, along with any imports defined in it?

(I'm referring to the use of `(new-interaction-environment)` and whether it consults `init.ss` as part of the default logic, whether or not a user supplies something other than the default `'(ironscheme)` libspec.)

**Reply by leppie** [reply]
I cant remember :D You will to test and see. And let me know if something is not working as expected.

**Comment by bcalco** [comment]
@leppie I need to query the interaction environment to determine if a specific symbol has been bound in it. How can I do this in IronSchemes (it seems different Schemes do it differently, if at all)? In general, I'd like to query environments for such bound identifiers.

I tried the following but got an error:

```
> (environment-bindings (interaction-environment))
Unhandled CLR exception during evaluation:
CLR Exception: System.InvalidCastException
System.InvalidCastException: Unable to cast object of type 'record.g$13251$IY5sc.interaction-env' to type 'record.g$13234$IY5sc.env'.
   at #.psyntax.expander::environment-bindings(Object e)
   at #.ironscheme.exceptions::dynamic-wind(Object in, Object proc, Object out)
   at #.ironscheme.exceptions::dynamic-wind(Object in, Object proc, Object out)
   at IronScheme.Runtime.Builtins.CallWithCurrentContinuation(Object fc1)
   at IronScheme.Runtime.R6RS.Exceptions.WithClrExceptionHandler(Object handler, Object thunk)

>
```

**Reply by bcalco** [reply]
And by the way if it's doable from the C# side, I'm cool to use that too. It's a tooling issue as I'm adding UX to my solution now and want to provide certain visuals, like "what symbols are defined in the current environment?" But bonus if I can do it from Scheme, too.

**Reply by leppie** [reply]
`environment-symbols` takes an environment as an argument.

Eg

```
> (environment-symbols (environment '(rnrs r5rs)))
(modulo remainder quotient scheme-report-environment null-environment delay exact->inexact force inexact->exact)
```

**Reply by leppie** [reply]
Even simpler `(interaction-environment-symbols)` :D 

I dont think `environment-bindings` works with the interaction environment as it is more to classify exports as procs, syntax or records.

**Reply by bcalco** [reply]
Thanks, I think once upon a time I knew that. LOL.

**Comment by leppie** [comment]
Newest release (309) has `EvalWithEnvironmentInstance` which should make managing environments from the caller's side a lot easier.

Sample usage:

```
// with this env needs to be extended, else define begin symbol-value may be unbound
var env = "(environment '(only (ironscheme) define begin symbol-value) '(rnrs))".Eval(); 
var ienv = "(interaction-environment)".Eval();
var newenv = "(new-interaction-environment)".Eval();

var r1 = "(+ 1 {0})".Eval(5);
var r2 = "(+ 1 {0})".EvalWithEnvironment("(environment '(rnrs))", 5);
var r3 = "(+ 1 {0})".EvalWithEnvironment("(interaction-environment)", 5);

var r4 = "(+ 1 {0})".EvalWithEnvironmentInstance(env, 5);
var r5 = "(+ 1 {0})".EvalWithEnvironmentInstance(ienv, 5);
var r6 = "(+ 1 {0})".EvalWithEnvironmentInstance(newenv, 5);
```
